### ISSUE TITLE
Add Custom Gravity Calculation in Area3D/Area2D

### ISSUE THREAD

**Issue by Remi123** [godotengine/godot-proposals#5714]
### Describe the project you are working on

A platformer similar to Super Mario Galaxy, with lots of custom Areas that define how gravity affect their bodies.

### Describe the problem or limitation you are having in your project

Custom gravity are restricted to an Area3D, where the shape only define if the custom gravity apply to them or not. However, the options for gravity vector can only represent directional ( default ) or spherical gravity (gravity_point is on). Calculations exists to get the gravity vector for cuboid, cone, curve and other shapes used in Super Mario Galaxy, but the current Area3D doesn't provide a virtual function to override the default behavior of direction or sphere. 

My first intuition was to inherit from Area3D and override a function to modify the gravity, since I want to use PhysicsServer3D.body_get_direct_state(get_rid()).total_gravity to retrieve it. But godot doesn't expose the 
void GodotArea3D::compute_gravity(const Vector3 &p_position, Vector3 &r_gravity) const function to the godot interface, and this function isn't virtual. 

Due to this, the only way I can add custom gravity is by moving the logic away from PhysicsServer with custom GetGravity function on each Area3D, and where each area is responsible to manage the bodies under its influence. Furthermore, properly mixing overlapping Areas is something that PhysicsServer is already managing. So this would just be reinventing the wheels.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Making virtual the function ```void GodotArea3D::compute_gravity(const Vector3 &p_position, Vector3 &r_gravity) const ``` and exposing it to classes that inherit from Area3Ds is the simplest option, but the default of directional or spherical gravity are really good and should stay the default, due to platformer games and space ship games with round planets. The interface of the Area3D Editor should concentrate on those two use-cases. 
 
The interaction with the interface and gravity_space_override, gravity_point_distance_scale, gravity_point_center, gravity_point, gravity_direction and gravity should be discussed but in my opinion, all except for gravity_space_override only serve the calculation of the gravity vector and can be kept as is for the user if it needs it. Gravity_space_override is used by the PhysicsServer3D when mixing multiple areas3d so it isn't really impacted by this.

The only problem with this solution is ABI compatibility, since the signature of the function changes to virtual 
 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

This is the classic Object Oriented example of Rectangle and Sphere classes both inheriting from Shapes class with the virtual GetArea() function, but we want to compute the direction of the gravity for an object in an Area3D.

I'm unfamiliar with Godot's code but I'm deducing that we can do something similar to RigidBody3D::_integrate_force(state), where this function is virtual and exposed to user if he needs to write it's own custom logic, and call it where appropriate with the macro GDVIRTUAL_CALL(_integrate_forces, p_state);, but instead it would be GDVIRTUAL_CALL_PTR(area,compute_gravity,position,result) where appropriate.

Note that the CollisionShape3D should still not influence the gravity vector, only where this is calculated. Of course I have plan to make my CylinderGravityArea3D extends Area3D to look into the cylinder's collisionshape3d's height and radius to calculate the gravity, but that's my own code that should affect no other.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Every article/tutorial I found online tackle this problem similarly : Bypass the current physics system and call a getGravity function from the current area. They either give the Body or the Area the responsibility to apply the gravity movement to the Body, but every PhysicsBody must inherit or have the component CustomGravity in order to get the non-physics gravity.

This means that using PhysicsServer3D.body_get_direct_state(get_rid()).total_gravity become a mistake, so if my scene have custom gravity areas, if I add a RigidBody3D without inheriting from this custom GravityBody, it won't be affected. Worse, we are essentially duplicating the work of the physics body. This prevent some optimizations like the parallelization of the calculation or letting other thread calculate it.

The alternative are possible, but I believe the integration with PhysicsServer3D is important and should at least be discussed.

### Is there a reason why this should be core and not an add-on in the asset library?

Implementing this in core allows for easier integration within the PhysicsServer calculation, allowing gravity to be reliable for all *Body3D, regardless if it's a character or a simple crate.

A single source of truth for something as universal as gravity is always a plus, even if the user have to provide the calculation for non-trivial cases. My solution is to offer a customization point while keeping the current default.

I open to other solutions

**Comment by addmix** [comment]
I agree entirely. This would be exceedingly useful for many projects, and can open the door, and streamline the execution, of making gravity like Super Mario Galaxy, or twisting walkways similar to Psychonauts 2. I have had the same difficulties, needing to create a custom `get_gravity()` function, which then doesn't replicate to all physics types. 

Might I suggest alongside this proposal, exposing the currently felt gravity force to the `PhysicsBody3D` class, to circumvent the need for this cumbersome line `PhysicsServer3D.body_get_direct_state(get_rid()).total_gravity`

**Comment by aaronfranke** [comment]
Ah, I forgot this proposal existed and ended up opening basically a duplicate of this: https://github.com/godotengine/godot-proposals/issues/7850

I already have a pull request opened with a proposed implementation. It adds a callback method on Area2D/3D that allows you to compute custom gravity based on the relative position of the object. This callback integrates with and adds to the existing `GodotArea3D::compute_gravity` method instead of replacing it.

Both proposals have the goal of allowing Super Mario Galaxy style gravity in Godot. Also note that I have a further goal of standardizing this gravity inside of glTF, so you can store entire levels including their gravity inside of a 3D model that is designed to be engine-agnostic. That is separate from the Godot implementation, but worth mentioning.
